onnxruntime/core/external_ops/pyop.cc

```diff
@@ -0,0 +1,219 @@
+private:
+    static std::mutex mtx_;
+    vector<PyObject*> objs_;
+};
+
+std::mutex Scope::mtx_;
+
+PYOP_EXPORT bool Initialize() 
+{
+    Scope scope;
+    Py_Initialize();
+    if (_import_array() < 0) {
+        return false;
+    }
+    auto path_list = PySys_GetObject("path");// do not release it
+    if (nullptr == path_list || !PyList_Check(path_list) ||
+        PyList_Append(path_list, PyUnicode_FromString(".")) != 0) {
+        return false; 
+    }
+    static Finalizer finalizer;
+    return true;
+}
+
+PYOP_EXPORT const char* GetLastErrorMessage(std::string& err)
+{
+    Scope scope;
+    if (PyErr_Occurred()) {
+        PyObject *type, *value, *trace;
+        PyErr_Fetch(&type, &value, &trace);
+        if (nullptr != value) {
+            auto pyVal = PyObject_Repr(value);
+            scope.Add(pyVal);
+            auto pyStr = PyUnicode_AsEncodedString(pyVal, "utf-8", "Error ~");
+            scope.Add(pyStr);
+            err = PyBytes_AS_STRING(pyStr);
+        }
+        PyErr_Restore(type, value, trace);
+    }
+    return err.c_str();
+}
+
+PyObject* MakePyObj (const void* data, int32_t type, const vector<int64_t>& dim)
+{
+    std::vector<npy_intp> np_dim;
+    for (auto d: dim) {
+        np_dim.push_back(static_cast<npy_intp>(d));
+    }
+    auto pyObj = static_cast<PyObject*>(PyArray_EMPTY(static_cast<int>(np_dim.size()), np_dim.data(), type, 0));
+    auto data_len = std::accumulate(begin(np_dim), end(np_dim),
+                                    static_cast<int64_t>(PyArray_DescrFromType(type)->elsize),
+                                    std::multiplies<int64_t>());
+    auto np_array = reinterpret_cast<PyArrayObject*>(pyObj);
+    memcpy(PyArray_DATA(np_array), data, data_len);
+    return pyObj;
+}
+
+bool ExtractOutput (PyObject*                pyObj,
+                    vector<const void*>&     outputs,
+                    vector<int32_t>&         outputs_elem_size,
+                    vector<vector<int64_t>>& outputs_dim)
+{
+    if (!PyArray_Check(pyObj)) {
+        return false;
+    }
+
+    outputs_dim.push_back({});
+    auto np_array = reinterpret_cast<PyArrayObject*>(pyObj);
+    outputs_elem_size.push_back(static_cast<int32_t>(PyArray_ITEMSIZE(np_array)));
+
+    for (int i = 0; i < PyArray_NDIM(np_array); ++i) {
+        outputs_dim.back().push_back(PyArray_SHAPE(np_array)[i]);
+    }
+
+    auto data_len = std::accumulate(begin(outputs_dim.back()),
+                                    end(outputs_dim.back()),
+                                    static_cast<int64_t>(outputs_elem_size.back()),
+                                    std::multiplies<int64_t>());
+
+    outputs.push_back(new char[data_len]);
+    memcpy(const_cast<void*>(outputs.back()), PyArray_DATA(np_array), data_len);
+    return true;
+}
+
+PYOP_EXPORT void* NewInstance (const char* module, const char* class_name, const unordered_map<string, string>& args)
+{
+    Scope scope; 
+    auto pyModule = PyImport_ImportModule(module);
```

Comment:
I think you should DECREF when leaving the function. Your code is not using a binding such as pybind11 so you need to take care of DECREF yourself. If missing, the impact should be noticed after many calls the operators with some memory leak.